src/Orleans/Core/Grain.cs

```diff
@@ -39,7 +36,7 @@ protected IGrainFactory GrainFactory
         /// </summary>
         protected IServiceProvider ServiceProvider 
         {
-            get { return Runtime?.ServiceProvider; }
+            get { return Data.ServiceProvider ?? Runtime?.ServiceProvider; }
```

Comment:
I don't think facets will work that way anyway, there will probably still be a singleton service that acts as a factory for the facets, and they will not be stored in the container by scoping it, otherwise you wouldn't be able to have 2 facets of the same type (but with different name) in the same grain.

BTW, MVC does allow a way to opt out of scoping per request...